framework/src/Volo.Abp.EntityFrameworkCore/Volo/Abp/EntityFrameworkCore/Modeling/AbpPropertyBuilderExtensions.cs

```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Linq.Expressions;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+using Newtonsoft.Json;
+
+namespace Volo.Abp.EntityFrameworkCore.Modeling
+{
+    public static class AbpPropertyBuilderExtensions
+    {
+        public static void ConfigureJsonConversionWithValueComparer<TPropertyType>(
+            this PropertyBuilder<TPropertyType> propertyBuilder,
+            Expression<Func<TPropertyType, TPropertyType, bool>> equalsExpression,
+            Expression<Func<TPropertyType, int>> hashCodeExpression,
+            Expression<Func<TPropertyType, TPropertyType>> snapshotExpression)
+        {
+            propertyBuilder.HasConversion(
+                    d => JsonConvert.SerializeObject(d, Formatting.None),
+                    s => JsonConvert.DeserializeObject<TPropertyType>(s)
+                )
+                .Metadata.SetValueComparer(new ValueComparer<TPropertyType>(
+                    equalsExpression,
+                    hashCodeExpression,
+                    snapshotExpression));
```

Comment:
Why pass these as parameters instead of implementing like `.HasConversion`? It's the same for all callers.

Reply:
I think this extension method may be called by developers, their entity properties may not be `Dictionary`. So the expression may be different.

Reply:
This implementation for `.HasConversion` already assumes the property is a `Dictionary` right?

Reply:
`ConfigureJsonConversionWithValueComparer `uses the generic parameter `TPropertyType`, which may be a `Dictionary `or a developer-defined `ICollection` etc. but I think I should add other extension methods to support non-collection or non-enumeration types.

eg `ConfigureJsonConversion `

Reply:
We shouldn't add other extension methods that are unlikely to be used.

We should have one specifically for `Dictionary` to extract our repeated and reusable code.

Reply:
@maliming Pending your changes.

Reply:
hi @acjh 
These methods (`ConfigureJsonConversionWithValueComparer`) can be used both inside the framework and by developers. 
The method does not require that the properties of the entity must be `Dictionary`,
 If the developer's properties are `string`, or `list`, etc. they can call `ConfigureJsonConversionWithValueComparer`.
 This does not Ok?

Reply:
We shouldn't add other extension methods that are unlikely to be used.
Consider the following.

Raw, 8 lines repeated 3 times in framework: straightforward, but most verbose

```c#
b.Property<Dictionary<string, object>>(nameof(IHasExtraProperties.ExtraProperties))
    .HasColumnName(nameof(IHasExtraProperties.ExtraProperties));
    .HasConversion(
        d => JsonConvert.SerializeObject(d, Formatting.None),
        s => JsonConvert.DeserializeObject<Dictionary<string, object>>(s)
    )
    .Metadata.SetValueComparer(new ValueComparer<TPropertyType>(
        (d1, d2) => d1.SequenceEqual(d2),
        d => d.Aggregate(0, (k, v) => HashCode.Combine(k, v.GetHashCode())),
        d => d.ToDictionary(k => k.Key, v => v.Value));
```

Option 1 (current), 5 lines repeated: partly generic, but still verbose

```c#
b.Property<Dictionary<string, object>>(nameof(IHasExtraProperties.ExtraProperties))
    .HasColumnName(nameof(IHasExtraProperties.ExtraProperties));
    .ConfigureJsonConversionWithValueComparer(
        (d1, d2) => d1.SequenceEqual(d2),
        d => d.Aggregate(0, (k, v) => HashCode.Combine(k, v.GetHashCode())),
        d => d.ToDictionary(k => k.Key, v => v.Value)
    );
```

Option 2: least repeated code, but least reusable — my previous proposal that you didn't (and I now don't) want

```c#
b.Property<Dictionary<string, object>>(nameof(IHasExtraProperties.ExtraProperties))
    .HasColumnName(nameof(IHasExtraProperties.ExtraProperties));
    .ConfigureJsonConversionWithValueComparerForDictionary();
```

Option 3: concise and most discoverable

```c#
b.Property<Dictionary<string, object>>(nameof(IHasExtraProperties.ExtraProperties))
    .HasColumnName(nameof(IHasExtraProperties.ExtraProperties));
    .HasConversionForJson()
    .MetaData.SetValueComparerForDictionary();
```

Option 4: less repeated code and most clean (in terms of convention and not polluting with extensions)

```c#
b.Property<Dictionary<string, object>>(nameof(IHasExtraProperties.ExtraProperties))
    .HasColumnName(nameof(IHasExtraProperties.ExtraProperties));
    .HasConversion(new AbpJsonValueConverter())
    .MetaData.SetValueComparer(new AbpDictionaryValueComparer());
```

We currently use Option 4 with `AbpDateTimeValueConverter`.
I'm inclined towards Option 4 now.

Reply:
@acjh 👍 
@hikalkan What do you think?

Reply:
@maliming Only for properties that are `collection or enumeration type`, as mentioned in that error message.

Corresponding API browser entry: https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.diagnostics.coreeventid.collectionwithoutcomparer?view=efcore-3.1

Reply:
@hikalkan Pending your thoughts.

Reply:
It is too long discussion for me 😄 But I will check when I have time.

Reply:
Thanks to both of you :)
Let's do it like that: @acjh you continue this issue, implement your thoughts and I will review it finally and merge. Please try to complete it before this thursday.

Reply:
@hikalkan @maliming I have implemented Option 4.

Reply:
Great! Thanks to both of you. I will review and merge it.